Libplanet/Net/Peer.cs

```diff
@@ -70,11 +85,39 @@ protected Peer(SerializationInfo info, StreamingContext context)
         [Pure]
         public IPAddress PublicIPAddress { get; }
 
+        public bool IsCompatibleWith(
+            AppProtocolVersion localVersion,
+            IEnumerable<PublicKey> trustedSigners = null,
+            DifferentAppProtocolVersionEncountered recognizer = null
+        )
+        {
+            if (AppProtocolVersion.Equals(localVersion))
+            {
+                return true;
+            }
+
+            if (trustedSigners is null || trustedSigners.Any(s => AppProtocolVersion.Verify(s)))
+            {
+                return recognizer is null
+                    ? false
+                    : recognizer(this, AppProtocolVersion, localVersion);
```

Comment:
It may be a little complicated to run the callback in this predicate... 😕

Reply:
Indeed, but I haven't found any better approach yet… 😇